### ISSUE TITLE
`cabal run` creates `dist-newstyle/cache/compiler` in the current folder

### ISSUE THREAD

**Issue by Bodigrim** [haskell/cabal#10686]
**Describe the bug**
Running a standalone script (`#!/usr/bin/env cabal`) creates and leaves `dist-newstyle/cache/compiler` in the current folder. This is a regression in comparison to Cabal 3.14.1.0.

**To Reproduce**

```
$ ll dist-newstyle/cache/compiler; echo '#!/usr/bin/env cabal\n{- cabal:\nbuild-depends: base\n-}\nmain :: IO ()\nmain = pure ()' > Main.hs && chmod +x Main.hs && ./Main.hs && ll dist-newstyle/cache/compiler
ls: dist-newstyle/cache/compiler: No such file or directory
-rw-r--r--  1 andrew  staff  25365 Dec 29 01:31 dist-newstyle/cache/compiler
```

**Expected behavior**
With Cabal 3.14.1.0 no files are created in the current folder.

**System information**
 - macOS, GHC 9.6.6, Cabal HEAD built from 4b2e5f03f

**Comment by Bodigrim** [comment]
IMHO this is a fairly grave regression, rendering Cabal scripts unusable. Could someone please reproduce and triage?

**Comment by ffaf1** [comment]
I confirm I can reproduce it.

If I modify the `.hs` and `dist-newstyle` is already present, I get a `build/` folder too.

```
f@x270:/tmp/prova$ ls
build  dist-newstyle  Main.hs
```

**Comment by ffaf1** [comment]
c89ab54e13f1eb808dcdd470edecd7eaaa594cec  #10254

@alt-romes does this ring a bell?

**Comment by alt-romes** [comment]
> c89ab54e13f1eb808dcdd470edecd7eaaa594cec  #10254
> 
> @alt-romes does this ring a bell?

It does not, sorry. Do you think the commit you referenced is related to this regression? Why?

**Comment by alt-romes** [comment]
Ah, actually some of those lines could be culpable for this:

https://github.com/haskell/cabal/commit/c89ab54e13f1eb808dcdd470edecd7eaaa594cec#diff-2adae381a2eb2d8c804d5396694a431e1e3babd816a5d2408a62213cf774a893L388-L398

The call to `createDirectory` was there before too, but it was not necessarily invoked (it depends on what `initprojectskeleton` did with it.

**Comment by ffaf1** [comment]
@alt-romes it definitely looks like it could be related to the problem.
I just did a `git bisect` and landed on that commit.

Are you working on it? If you need an additional pair of eyes to diagonse this, I can gladly help!

**Comment by alt-romes** [comment]
I'll take a look today.

**Comment by alt-romes** [comment]
Fixed in #10771

Note: the `build` directory that is also created is caused by some earlier commit. The `build` directory is still created after #10771. This patch only fixes the incorrectly created `dist-newstyle`.

@ffaf1 perhaps it would be good to create a new ticket for the `build` directory being created

**Comment by ffaf1** [comment]
Thanks and

> @ffaf1 perhaps it would be good to create a new ticket for the build directory being created

done!
